fvm/Equation/MomentQuantity.cpp

```diff
@@ -4,60 +4,198 @@
 
 using namespace DREAM::FVM;
 
 
 /**
  * Constructor.
  */
-MomentQuantity::MomentQuantity(Grid *momentGrid, Grid *fGrid, len_t momentId, len_t fId) 
-    : EquationTerm(momentGrid), fGrid(fGrid), momentId(momentId), fId(fId) {
-    
-    this->GridRebuilt();
+MomentQuantity::MomentQuantity(Grid *momentGrid, Grid *fGrid, len_t momentId, len_t fId, UnknownQuantityHandler *u, real_t pThreshold, pThresholdMode pMode) 
+    : EquationTerm(momentGrid), fGrid(fGrid), momentId(momentId), 
+      fId(fId), unknowns(u), pThreshold(pThreshold), pMode(pMode) {
+    this->hasThreshold = (pThreshold!=0);
+
+    if ((pMode == P_THRESHOLD_MODE_MIN_THERMAL_SMOOTH) || (pMode == P_THRESHOLD_MODE_MAX_THERMAL_SMOOTH)){
+        AddUnknownForJacobian(u->GetUnknownID(OptionConstants::UQTY_T_COLD));
+        smoothEnvelopeStepWidth = 2;
+    }
+    this->GridRebuilt();    
 }
 
 /**
  * Destructor.
  */
 MomentQuantity::~MomentQuantity() {
     delete [] this->integrand;
+    delete [] this->diffIntegrand;
 }
 
 /**
- * Method that is called whenever the grid is rebuilt.
- * Here, we use it to rebuild the 'integrand' variable.
+ * Rebuild allocates memory for the integrand and diffIntegrand
+ * arrays whenever the grid has changed size.
  */
 bool MomentQuantity::GridRebuilt() {
     bool rebuilt = this->EquationTerm::GridRebuilt();
-    
     const len_t N = this->fGrid->GetNCells();
-
     if (this->nIntegrand != N) {
         this->nIntegrand = N;
+        if(integrand != nullptr)
+            delete [] integrand;
         this->integrand = new real_t[N];
 
-        // Figure out the maximum number of non-zeros needed per
-        // matrix row...
-        const len_t nr = fGrid->GetNr();
-        this->nnz_per_row = 0;
-        for (len_t i = 0; i < nr; i++) {
-            len_t nc = fGrid->GetMomentumGrid(i)->GetNCells();
-            if (this->nnz_per_row < nc)
-                this->nnz_per_row = nc;
-        }
+        if(MaxNMultiple())
+            AllocateDiffIntegrand();
 
         return true;
-    } else return rebuilt;
+    } else
+        return rebuilt;   
 }
 
+void MomentQuantity::AllocateDiffIntegrand(){
+    len_t nMultiples = MaxNMultiple();
+    if(nMultiples){
+        if(diffIntegrand != nullptr)
+            delete [] diffIntegrand;    
+        const len_t N = this->fGrid->GetNCells();
+        this->diffIntegrand = new real_t[nMultiples*N];
+    }
+}
+
+
 
 /**
- * Set the jacobian elements for this term. This assumes that
- * the integrand does not depend on any unknown quantity (other than
- * linearly on the unknown to which this moment operator is applied
- * (usually the distribution function f)).
+ * returns the momentum grid spacing dp at momentum p0
+ */
+real_t FindThresholdStep(real_t p0, MomentumGrid *mg){
+    const real_t *p = mg->GetP1();
+    for(len_t i=0; i<mg->GetNp1(); i++)
+        if(p[i]>p0)
+            return mg->GetDp1(i);
+    return p0 - mg->GetP1(mg->GetNp1()-1); // what else to do if p0 is outside the grid? 
```

Comment:
I don't know what the optimal solution is here, but if it is maybe we should remove the comment?

Reply:
This is due for a slight rework, as Ida identified a case where the method can lead to undesired behavior.